vpr/src/analytical_place/partial_legalizer.cpp

```diff
@@ -1587,72 +1669,158 @@ void BiPartitioningPartialLegalizer::partition_blocks_in_window(
+        lower_window_utilization -= blk_mass;
+        lower_window_overfill = std::move(new_lower_overfill);
+        upper_window_utilization = std::move(new_upper_utilization);
+        upper_window_overfill = std::move(new_upper_overfill);
+
+        // Add this block to the list of blocks to move and invalidate the block
+        // in the current list.
+        // Here, we mark the block as invalid and later we will compress the vector.
+        lower_blocks_to_move.push_back(lower_blk);
+        lower_contained_blocks[i] = APBlockId::INVALID();
+    }
+
+    // Try to move things from upper to lower greedily.
+    std::vector<APBlockId> upper_blocks_to_move;
+    upper_blocks_to_move.reserve(upper_contained_blocks.size());
+    for (size_t i = 0; i < upper_contained_blocks.size(); i++) {
+        // If the upper window has no overfill, there is no reason to move anything.
+        if (upper_window_overfill.is_zero())
+            break;
+
+        // Check what would be the gain of moving this block to the upper partition.
+        APBlockId upper_blk = upper_contained_blocks[i];
+        const PrimitiveVector& blk_mass = density_manager_->mass_calculator().get_block_mass(upper_blk);
+
+        PrimitiveVector new_upper_overfill = upper_window_overfill - blk_mass;
+        new_upper_overfill.relu();
+
+        float upper_gain = upper_window_overfill.manhattan_norm() - new_upper_overfill.manhattan_norm();
+        // If there is no gain on removing this from the upper partition, keep it where it is.
+        if (upper_gain == 0.0)
+            continue;
+
+        // Compute the change in overfill on the lower.
+        PrimitiveVector new_lower_utilization = lower_window_utilization + blk_mass;
+        PrimitiveVector new_lower_overfill = new_lower_utilization - lower_window_capacity;
+        new_lower_overfill.relu();
+
+        float lower_loss = new_lower_overfill.manhattan_norm() - lower_window_overfill.manhattan_norm();
+        // If we would lose more (i.e. make the overfill worse) by moving to upper than we would gain, keep on this side.
+        if (lower_loss >= upper_gain)
+            continue;
+
+        // Move this block to the other side.
+        upper_window_utilization -= blk_mass;
+        upper_window_overfill = std::move(new_upper_overfill);
+        lower_window_utilization = std::move(new_lower_utilization);
+        lower_window_overfill = std::move(new_lower_overfill);
+
+        upper_blocks_to_move.push_back(upper_blk);
+        upper_contained_blocks[i] = APBlockId::INVALID();
+    }
+
+    // Move the blocks into the windows.
+    //  Here we maintain the order the best we can to make future sorts faster.
+    // Compress the lower contained blocks which have not moved.
+    lower_window.contained_blocks.reserve(lower_contained_blocks.size() - lower_blocks_to_move.size() + upper_blocks_to_move.size());
+    for (APBlockId lower_blk : lower_contained_blocks) {
+        if (!lower_blk.is_valid())
```

Comment:
Would be a bit shorter and more clear to reverse this to 
if (lower_blk.is_valid()) {
   lower_window.contained ...
}